Sdtrig.tex

```diff
@@ -275,6 +275,36 @@ \subsection{Combined Accesses}
 loads of size SEW (selected element width), and {\tt cm.push} should be treated
 as if it performed multiple stores of size XLEN.
 
+\subsection{Cache Operations}
+
+\begin{steps}{
+Cache operations are unusual events, and code that uses them can have
+hard-to-find bugs. For the purposes of debug triggers, two classes of cache
+operations match as stores against any of the affected addresses:}
+\item Cache operations that enable software to maintain coherence between
+otherwise non-coherent implicit and explicit memory accesses.
+\item Cache operations that perform block writes of constant data.
+\end{steps}
+The access size for such operations is unknown to the TM, so only triggers with
+\FcsrMcontrolSixSize=0 will match.
```

Comment:
It's not clear to me that the size is always unknown to the TM.  The recommendation in select=0 (to compare on all addresses that are affected by an operation) mean that you would ideally be comparing on all 64 addresses (or however long the cache block size is) which means that a recommended implementation actually would know the size.

This also brings up a question.  Consider the case where we match on store address and tdata2=4.  If you do a cache op with rs1=4 then is this required to match?  My question has to do with "the lowest virtual address of the access" in the select=0 text.  If you only have one comparison then does that comparison need to clear the low 6 bits before comparing (since the lowest VA of the access is 0) or are you allowed to match on the address provided in the register?  The former would not match if rs1=4 even though tdata2=4 but the latter would (and would not match on the lowest VA). 
 This is about a minimally compliant implementation, not one that complies with the recommendation.

Reply:
I think that "is unknown to the TM" is just wordsmithing which can maybe be done post-freeze.  The other question about lowest VA is probably more substantive.

My inclination is to change "lowest virtual address of the access" in select=0 to say "lowest virtual address of the access or in the case of cache operations the base address from the register."  Basically, it's always the output of the base+offset calculation.  Then nothing in this section needs to change (other than minor clarifications).

Reply:
> It's not clear to me that the size is always unknown to the TM. The recommendation in select=0 (to compare on all addresses that are affected by an operation) mean that you would ideally be comparing on all 64 addresses (or however long the cache block size is) which means that a recommended implementation actually would know the size.

This is referring to the *access* size, not the number of bytes affected by the operation. The number of bytes should be known.

Reply:
I can see arguments either way.  It's a little awkward to add a masking capability before the comparison just for this one case.

However, the CMO spec says:
> the compare values
follow from the addresses of the NAPOT memory region corresponding to the cache block containing
the effective address

To me, that means that the CMO group was aware of this concept of multiple compare values, that the lowest VA of the line is included in that set of compare values, and that they didn't say anything along the lines of my proposal.

I'll drop my question and let's go with the status quo of lowest VA.

My only (minor) concern now is "The access size for such operations is unknown to the TM" which just doesn't seem necessarily true.

Reply:
> I can see arguments either way. It's a little awkward to add a masking capability before the comparison just for this one case.

Don't you already do that masking to find the correct cache line? On the other hand, maybe the cache is doing the masking, and the instruction execution unit doesn't actually know the cache line size or anything. In which case all it can do is go by the value given to the cache.

So... should (but not must) match against the lowest address in the cache line?

Reply:
I'm OK with should.